### PR TITLE
Change constrain to accept any boolean expression

### PR DESCRIPTION
**Pull request by jfecher**
Implements #235 

Changes the syntax for `constrain` from `constrain <expr> compare_op <expr>` to `constrain <expr>` where the now-single expression can be any boolean value.

The lack of duplicated handling of constrain operators lead to a lot of deletion of duplicate code and some bug fixes in type checking regarding this duplicate code often handling the various comparison operators differently when they're in a constrain statement versus outside of them.

There have been some ssa tests that rely upon `==` being implemented for arrays which was previously only possible inside of a `constrain`. The type checker has been slightly changed to allow this for all uses of `==` to maintain functionality.

This change caught some odd cases in tests which failed to be caught previously, including some comparisons of `Field == u8` and even `[]u8 == u8` in `examples/11/src/main.nr`.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/acir_gen.rs
Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:138]: @guipublic could you take a look at this?

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:138]: Will do after the merge of #221, I don't want to merge again and again.

Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:138]: Lets not hold up other PRs waiting for others to merge since its unclear how long until #221 will be merged itself. I can help with that merge as well if you'd like

**Review by guipublic — CHANGES_REQUESTED** [review]
The problem I see with this approach is that it introduces a slight overhead as we now requires to compute the boolean expression and checks its value is 1, while before we where evaluating the boolean expression in the constrain context, so only checking if it is different than 0.
What we can easily do to support any boolean expression but keep the special treatment for Eq and Neq, is when we parse the constrain expression in code_gen, if we do not match a boolean operator that we support, we generate `constrain e == 1`, where e is the boolean expression.

**Comment by jfecher** [comment]
Is there an even more direct translation we can do? For example, what is the difference between `my_bool` and `constrain my_bool` in acir, assuming neither are optimized out.

**Comment by guipublic** [comment]
> what is the difference between `my_bool` and `constrain my_bool` in acir, assuming neither are optimized out.
`constrain my_bool == 1` will output `my_bool = 1`, which cannot be optimized more.
`my_bool` in itself should just be a witness.
The small optimisation we get is only for the != constraint, although the == constraint for arrays could be also improved in future if we can use randomness (but this is another story)

**Comment by jfecher** [comment]
Hmm, my question was more grasping at the question of what a constraint is inherently rather than say `constrain x == 1` specifically. So that we can reuse the work already done on x when going from x to constrain x.
Small note that this PR does not actually do `constrain x == 1` for constraining x, it is meant to lower to `constrain x != 0` in the acir_gen pass.

**Comment by guipublic** [comment]
If you go from `x` to `constrain x`, you can re-use all the work done for x. By the way I noticed what you did for` constrain x` and it is not necessary, you can directly use `constrain x == 1`
My point was that when we have `x= a<>b`, in that case doing `x` is doing a bit more than `constrain x`. You can still re-use the work done for x, but in that case you do more than what is strictly required.
More precisely, `x` is doing:
```
y=x.m;
x.y=x;
```
While for constrain x, we just need
`x.m=1`

In the first case, it can return 0 or one (=y), while in the second case, it must be one and we save one multiplication.

**Comment by jfecher** [comment]
Hmm, what should be done in acir gen to fix this?

**Comment by guipublic** [comment]
> what should be done in acir gen to fix this?

We would require an optimisation pass which is able to optimise this. In this case it would be some kind of constant folding for acir_gen; when we write `y=1`, we replace `y` by 1 in all gates that use it, and the `x*y=x `would become `x=x` which would be removed.
The problem is that the number of gates can become very high (millions of gates), so doing a linear scan on all gates for every constant folding is not practical, so we need to be clever.
Just thinking out loud, we could write the 'constant gates' at the beginning (using a double queue) and then do a copy propagation-like pass to replace all the constants in one pass, and every updated gate is checked for simplification.

Alternatively, as I suggested, we can write directly the optimal version! (n.b. but this works only if the constrain is written with the operator `constrain x<>0,` not if written `y=x<>0; constrain y;`)

**Comment by guipublic** [comment]
I was thinking about this and I think we could do something like that:
When we constrain an expression, if the expresssion is a Neq instruction (to be checked but I think it is only this case that has an impact) , we mark it somehow as 'for a constraint statement'.
This mark will do nothing, except in acir_gen and also except that the mark must be kept through flatenning and inlining, and lost when the instruction is optimised during CSE.
Acir_gen would not process such marked instructions, but indirectly when processing the constraint instruction.
That way we keep all your implementation, except for acir_gen where we keep the previous behavior.

What do you think?
If it is ok for you, I can validate this PR and the change I suggest here can be done in another PR.

**Comment by jfecher** [comment]
That sounds quite good, it'd let ssa and previous passes mostly re-use the existing machinary for working on booleans for constraints while having only the acir_gen pass which actually cares about the difference use it.

And yes, if you could handle your idea in a separate PR that'd be good. We could then merge it into this PR or merge this one first then the second, either works for me.

File: crates/noirc_evaluator/src/ssa/acir_gen.rs
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:134]: There is no need to do this, the value is already 0 or 1 so we only need to check that it is one by doing something like this:
` evaluator.gates.push(subtract(&Arithmetic::one(), FieldElement::one(), &value ).into());`

**Review by guipublic — CHANGES_REQUESTED** [review]
Ok let's do this then!
We merge this PR first (once the comment is resolved), I will do the 2nd part later on, it is not urgent as it is a very small optimisation.

File: crates/noirc_evaluator/src/ssa/acir_gen.rs
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:134]: Alright, removed. I also returned `value` from the match, but perhaps I should early return from the whole function or return something else from the match instead. Not sure what the implications would be.

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs]: A constrain does not have a value, so you could return Arithmetic::default() instead.
Whatever we return will not be used and should not matter.

**Review by guipublic — DISMISSED** [review]

**Review by guipublic — APPROVED** [review]

**Pull request merged**
